### PR TITLE
SWTCH 1098 refactor and create unit tests for addrole

### PR DESCRIPTION
**Pull request by amaugosomto**

### CODE REVIEW

File: src/app/routes/applications/new-role/components/role-field/role-field.component.ts [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+  templateUrl: './role-field.component.html',
+  styleUrls: ['./role-field.component.scss']
+})
+export class RoleFieldComponent {
+
+  @Input() fieldsForm: FormGroup;
+  @Input() dataSource: MatTableDataSource<any>;
+  @Input() isChecking: boolean;
+
+  @Output() resetFormEvent = new EventEmitter<boolean>();
```
Comment (DawidGil-jlabs) [inline src/app/routes/applications/new-role/components/role-field/role-field.component.ts:20]: Do we need `Event` in property name?

File: src/app/routes/applications/new-role/components/role-field/role-field.component.ts [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+  @Output() updateDataSourceEvent: EventEmitter<MatTableDataSource<any>> = new EventEmitter();
+  @Output() backEvent: EventEmitter<any> = new EventEmitter();
+  @Output() proceedConfirmDetailsEvent: EventEmitter<any> = new EventEmitter();
```
Comment (DawidGil-jlabs) [inline src/app/routes/applications/new-role/components/role-field/role-field.component.ts:23]: Can we name it `confirm` / `next` / `proceed` or something like this?

File: src/app/routes/applications/new-role/components/role-field/role-field.component.spec.ts [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+      spyOn(component, 'resetForm');
+
+      component.cancelAddField();
+
+      expect(component.resetForm).toHaveBeenCalled();
+    });
+
+    it('moveDown', () => {
+      component.dataSource.data = [1, 2, 3];
+      const i = 1;
```
Comment (DawidGil-jlabs) [inline src/app/routes/applications/new-role/components/role-field/role-field.component.spec.ts:225]: this variable is not needed i think. Or if you want it, then please name it better. For example `const indexToMove = 1`

File: src/app/routes/applications/new-role/new-role.component.spec.ts [hunk 4]
```diff
@@ -1,25 +1,259 @@
...
+    'getDid',
+    'checkExistenceOfDomain',
+    'isOwner',
+    'getRoleDIDs',
+    'createRole',
+    'setRoleDefinition',
+    'getENSTypesBySearchPhrase'
+  ]);
+  const iamServiceSpy = jasmine.createSpyObj('iam', [
+    'isAlphaNumericOnly',
```
Comment (DawidGil-jlabs) [inline src/app/routes/applications/new-role/new-role.component.spec.ts:36]: This method in iamservice is deprecated. Please use function from `utils/functions/is-alpha-numeric`. That way we could remove this spy. 
And i know it is the same code, but i started removing parts in IamService that have nothing to do with IAM object from iam-client-lib.

File: src/app/routes/applications/new-role/new-role.component.spec.ts [hunk 5]
```diff
@@ -1,25 +1,259 @@
...
+
+    it('dataSourceChangeHandler', () => {
+      const data = ['test value'];
+
+      component.dataSourceChangeHandler(data);
+
+      expect(component.dataSource.data).toEqual(data);
+    });
+
+    it('ngAfterViewInit', () => {
```
Comment (DawidGil-jlabs) [inline src/app/routes/applications/new-role/new-role.component.spec.ts:97]: This test just checking if method have been called but do not test behavior of the component. Imo you can delete it as well.

File: src/app/routes/applications/new-role/new-role.component.spec.ts [hunk 6]
```diff
@@ -1,25 +1,259 @@
...
+
+    it('formResetHandler', () => {
+      spyOn(component.fieldsForm, 'reset');
+
+      component.formResetHandler();
+
+      expect(component.fieldsForm.reset).toHaveBeenCalled();
+    });
+
+    it('dataSourceChangeHandler', () => {
```
Comment (DawidGil-jlabs) [inline src/app/routes/applications/new-role/new-role.component.spec.ts:89]: I'm not sure of the value of this test.

File: src/app/routes/applications/new-role/new-role.component.spec.ts [hunk 7]
```diff
@@ -1,25 +1,259 @@
...
 
-xdescribe('NewRoleComponent', () => {
+import { NewRoleComponent, RoleType } from './new-role.component';
+import { IamService } from '../../../shared/services/iam.service';
+import { ToastrService } from 'ngx-toastr';
+import { MAT_DIALOG_DATA, MatDialog, MatDialogRef } from '@angular/material/dialog';
+import { NO_ERRORS_SCHEMA } from '@angular/core';
+import { PreconditionTypes } from 'iam-client-lib';
+
+describe('NewRoleComponent', () => {
```
Comment (DawidGil-jlabs) [inline src/app/routes/applications/new-role/new-role.component.spec.ts:13]: 👍 for making this spec file working.

File: src/app/routes/applications/new-role/new-role.component.spec.ts [hunk 8]
```diff
@@ -1,25 +1,259 @@
...
+    });
+
+    it('ngAfterViewInit', () => {
+      spyOn(component, 'confirmParentNamespace');
+      component.ngAfterViewInit();
+      component.fieldsForm.setErrors({ incorrect: true });
+      expect(component.confirmParentNamespace).toHaveBeenCalled();
+    });
+
+    it('controlHasError', () => {
```
Comment (DawidGil-jlabs) [inline src/app/routes/applications/new-role/new-role.component.spec.ts:104]: I do not see value of this test also. You are testing here form group object from angular. I think setting this form should be in role-form.
The same with isRoleNameInValid test.

**Review by DawidGil-jlabs — COMMENTED** [review]
Some of the comments regarding test could be applied to every test like better test description.
Some test could be tested through clicking buttons in html.
There are some comments to the code. Let me know what do you think about them!

File: src/app/routes/applications/new-role/components/role-field/role-field.component.ts [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+
+  showFieldsForm = false;
+  isEditFieldForm = false;
+  fieldIndex: number;
+  public FieldTypes = FIELD_TYPES;
+
+  constructor(private changeDetectorRef: ChangeDetectorRef) { }
```
[the remark this answers is not in the document]
Comment (VOSpdUkraine) [reply src/app/routes/applications/new-role/components/role-field/role-field.component.ts]: updated

File: src/app/routes/applications/new-role/components/role-field/role-field.component.ts (hunk 2 above)
Comment (VOSpdUkraine) [reply src/app/routes/applications/new-role/components/role-field/role-field.component.ts]: updated

File: src/app/routes/applications/new-role/components/role-field/role-field.component.ts [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+})
+export class RoleFieldComponent {
+
+  @Input() fieldsForm: FormGroup;
+  @Input() dataSource: MatTableDataSource<any>;
+  @Input() isChecking: boolean;
+
+  @Output() resetFormEvent = new EventEmitter<boolean>();
+  @Output() updateDataSourceEvent: EventEmitter<MatTableDataSource<any>> = new EventEmitter();
+  @Output() backEvent: EventEmitter<any> = new EventEmitter();
```
[the remark this answers is not in the document]
Comment (VOSpdUkraine) [reply src/app/routes/applications/new-role/components/role-field/role-field.component.ts]: updated

File: src/app/routes/applications/new-role/components/role-field/role-field.component.ts (hunk 1 above)
Comment (VOSpdUkraine) [reply src/app/routes/applications/new-role/components/role-field/role-field.component.ts]: updated

File: src/app/routes/applications/new-role/components/role-field/role-field.component.spec.ts (hunk 3 above)
Comment (VOSpdUkraine) [reply src/app/routes/applications/new-role/components/role-field/role-field.component.spec.ts]: updated

File: src/app/routes/applications/new-role/components/role-field/role-field.component.spec.ts [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+
+      expect(component.proceedConfirmDetailsEvent.emit).toHaveBeenCalled();
+    });
+
+    it('resetForm', () => {
+      component.isEditFieldForm = true;
+      component.showFieldsForm = true;
+      spyOn(component.resetFormEvent, 'emit');
+
+      component.resetForm();
```
[the remark this answers is not in the document]
Comment (VOSpdUkraine) [reply src/app/routes/applications/new-role/components/role-field/role-field.component.spec.ts]: updated

File: src/app/routes/applications/new-role/components/role-field/role-field.component.spec.ts [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+    });
+
+    it('back', () => {
+      spyOn(component.backEvent, 'emit');
+      component.back();
+
+      expect(component.backEvent.emit).toHaveBeenCalled();
+    });
+
+    it('proceedConfirmDetails', () => {
```
[the remark this answers is not in the document]
Comment (VOSpdUkraine) [reply src/app/routes/applications/new-role/components/role-field/role-field.component.spec.ts]: updated

File: src/app/routes/applications/new-role/components/role-field/role-field.component.spec.ts [hunk 13]
```diff
@@ -0,0 +1,319 @@
...
+      component.fieldsForm.setErrors({ incorrect: true });
+      spyOn(component, 'resetForm');
+      spyOn(component, 'updateDataSource');
+      component.addField();
+
+      expect(component.resetForm).not.toHaveBeenCalled();
+      expect(component.updateDataSource).not.toHaveBeenCalled();
+    });
+
+    it('back', () => {
```
[the remark this answers is not in the document]
Comment (VOSpdUkraine) [reply src/app/routes/applications/new-role/components/role-field/role-field.component.spec.ts]: updated

File: src/app/routes/applications/new-role/components/role-field/role-field.component.spec.ts [hunk 14]
```diff
@@ -0,0 +1,319 @@
...
+
+      const field = component.dataSource.data;
+      expect(field[0].fieldType).toBe('number');
+      expect(field[0].label).toBe('Edit Label');
+      expect(component.showFieldsForm).toBeFalsy();
+    });
+
+  });
+
+  describe('should be run', () => {
```
[the remark this answers is not in the document]
Comment (VOSpdUkraine) [reply src/app/routes/applications/new-role/components/role-field/role-field.component.spec.ts]: updated

**Review by DawidGil-jlabs — APPROVED** [review]
Looks good.

**Pull request merged**